### PR TITLE
Building my app

### PR DESCRIPTION
**Pull request by mohamedgamaleldin999999**
In this PR, I am trying to build my capstone project following these requirements:

First you need to choose an API to base the development of the webapp on. The API should allow you to get numeric values based on a parameter. We recommend that you choose an API that is open or needs an API key. If you choose an API that require authentication, you should implement it on your own.

Design

You should follow these [design guidelines](https://www.behance.net/gallery/31579789/Ballhead-App-(Free-PSDs)), including:
Colors (select one main color for your website).
Typography: font face, size and weight.
Layout: composition and space between elements.
Original design idea by [Nelson Sakwa on Behance](https://www.behance.net/sakwadesignstudio).

The [Creative Commons license of the design](https://creativecommons.org/licenses/by-nc/4.0/) requires that you give appropriate credit to the author. Therefore, you must do it in the README of your project.

Interactions

Home page

When the page loads, the webapp shows the list of categories that could be filtered out by some parameter, for example by the category name.
Along with the category name, you will display some numeric values per category that come from the API.
When the user clicks (or taps) on a category item, the application navigates to the details page.
Details page

In the details page, the webapp retrieves data from the API to show detailed data in the category.
When the user clicks on the "Back" button (<), the user navigates to the home page.
Testing requirements

Create unit tests for pure functions in your Redux code.
Create integration tests for your application using the React Testing Library.
You may need to mock the access to the API, so that your tests don't send actual requests.
You may need to mock the connection to the Redux Store.
Technical requirements

The project is a single-page application (SPA) built with React and Redux.
The data retrieved from the API should be stored in the Redux store.
You should filter the data that you retrieve from the API using a Filter stateless component.
Every page (the main page and the pages for each item) should have a unique route within the SPA.
The project should be deployed and accessible online using [Render](https://render.com/), [Netlify](https://www.netlify.com/) or [GitHub Pages](https://pages.github.com/).
Add a link to the online version of your app to your README file.

### CODE REVIEW

File: src/App.test.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { render, screen } from '@testing-library/react';
+import App from './App';
+
+test('renders learn react link', () => {
+  render(<App />);
+  const linkElement = screen.getByText(/learn react/i);
+  expect(linkElement).toBeInTheDocument();
+});
```
Comment (AbayomiOlaoye) [inline src/App.test.js]: - [ ] This and 5 other tests are failing....Kindly make the tests run successfully to guarantee the considerable integrity or reliability of your code. :pray: 

image

- [ ] Kindly implement the filter functionality to smoothen user experience on your app according to the requirement shown below:
image

**Review by AbayomiOlaoye — CHANGES_REQUESTED** [review]
## STATUS: CHANGE REQUIRED ♻️ 

Hi @mohamedgamaleldin999999 , :wave: :wave: 
 

Good job so far!
There are some issues that you still need to work on to prepare your project for the final evaluation but you are almost there!

so_close

## Highlights
- Descriptive Pull Request message :heavy_check_mark: 
- Good use of GitHub flow :heavy_check_mark: 
- Meaningful commit messages :heavy_check_mark: 
- No linter error :heavy_check_mark: 
- Clean code :heavy_check_mark: 
- Fully functional and highly performant features :heavy_check_mark: 
- I love the preciseness and content of your video :+1: 
- Some tests are passing :heavy_check_mark: 
- Descriptive ReadMe documentation :heavy_check_mark: 


## Suggested changes
- [ ] _Check the comments under the review._

You can use as many of my suggestions as you want. If there is anything you would like to skip - feel free to do that. However, I strongly recommend you to take them into account as they can make your code better._

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

File: src/__tests__/Cities.test.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import React from 'react';
+import { Provider } from 'react-redux';
+import Adapter from 'enzyme-adapter-react-16';
```
Comment (tugsem) [inline src/__tests__/Cities.test.js]: - [ ] Please fix this error as the previous reviewer requested:
image

File: src/__tests__/CityList.test.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
+import { Provider } from 'react-redux';
+import Adapter from 'enzyme-adapter-react-16';
```
Comment (tugsem) [inline src/__tests__/CityList.test.js]: - [ ] Please fix this error as the previous reviewer requested: 
image

File: src/__tests__/Error.test.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import renderer from 'react-test-renderer';
```
Comment (tugsem) [inline src/__tests__/Error.test.js:2]: - [ ] Please fix this error as the previous reviewer requested: 
image

File: src/__tests__/Nav.test.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+import React from 'react';
+import renderer from 'react-test-renderer';
```
Comment (tugsem) [inline src/__tests__/Nav.test.js:2]: - [ ] Please fix this error as the previous reviewer requested:
image

File: src/__tests__/City.test.js [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import React from 'react';
+import { Provider } from 'react-redux';
+import Adapter from 'enzyme-adapter-react-16';
```
Comment (tugsem) [inline src/__tests__/City.test.js]: - [ ] Please fix this error as the previous reviewer requested:
image

**Review by tugsem — CHANGES_REQUESTED** [review]
## STATUS: REQUIRED CHANGES ♻️
Hello  @mohamedgamaleldin999999 👋 , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are some issues that you still need to work on to go to the next project but you are almost there!

Alt Text

 ## Highlights 🚀
✔️Good job on adding a filter method on the main page

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
- [ ] The horizontal scroll doesn't function. Please enable scrolling bar for the best user experience.
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/__tests__/City.test.js (hunk 6 above)
Comment (mohamedgamaleldin999999) [reply src/__tests__/City.test.js]: @tugsem thanks my friend.

the tests are passing on my local machine. Can you help me on why I am facing this problem?

File: src/index.js [hunk 7]
```diff
@@ -2,15 +2,15 @@ import React from 'react';
...
 import App from './App';
 import reportWebVitals from './reportWebVitals';
 
 const root = ReactDOM.createRoot(document.getElementById('root'));
 root.render(
   <React.StrictMode>
     <Provider store={store}>
-      <app />
+      <App />
     </Provider>
```
Comment (TadesseDev) [inline src/index.js]: - [ ] Dear @mohamedgamaleldin999999, I noticed you have made some changes to your project. Yet, The tests are still falling. This has to do with some packages/modules that are not available inside the test environments. As stated in the test results, you are missing `module 'enzyme-adapter-react-16' from 'src/__tests__/City.test.js'`, `module 'enzyme-adapter-react-16' from 'src/__tests__/CityList.test.js'`, `module 'react-test-renderer' from 'src/__tests__/Nav.test.js'` and `module 'enzyme-adapter-react-16' from 'src/__tests__/Cities.test.js'`, Please work on this tests and resubmit the project for another review. 

**In the case the test is running well on your computer, Please use a separate machine, clone the project, and try it there.**

Screen Shot 2023-09-09 at 9 56 21 PM

**Review by TadesseDev — CHANGES_REQUESTED** [review]
### Hi @mohamedgamaleldin999999 

### I'm Tadesse Alemayehu your reviewer for this PR. :hand: 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

199733072-bdacca19-8df9-43a7-be93-3aff2f1c15c8

 ## Highlights
 - [x] Good job working on the required changes 🙏


 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

- N/A


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by Efiamotu-1 — APPROVED** [review]
Hi @mohamedgamaleldin999999 , 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# STATUS: APPROVED 🕺🥳 

image

## Highlights :dart: 
* Good commit messages :heavy_check_mark: 
* All linters passing :heavy_check_mark: 
* Good PR summary :heavy_check_mark: 

## Optional suggestions
_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, you can now Merge since it is Approved ✅, However if there are any conflicts please resolve them before merging.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**